### PR TITLE
Implement simple vehicle model

### LINKED ISSUE #2940: Translational driving resistance
It would be nice to have a translational driving resistance at hand, containing drag resistance (dependent on square of speed), rolling resistance and gravitational resistance (depending on inclination).

### PR DESCRIPTION
**Pull request by AHaumer**
implement simple vehicle model (resolve #2940 / alternative PR to #2954)

### CODE REVIEW

**Review by christiankral — COMMENTED** [review]
Looks OK

**Review by christiankral — APPROVED** [review]

**Comment by AHaumer** [comment]
@tobolar we kindly ask you to review, seems like a mature proposal.

**Comment by beutlich** [comment]
@AHaumer Please not that the rebased on force-pushed the branch. Please reset on your side to get these changes.

**Review by beutlich — DISMISSED** [review]
The Vehicle examples mentions vehicle1D1/vehicle1D2 which are not available in the model.

**Comment by beutlich** [comment]
I believe there will be some general discussion if such a Vehicle model belongs to Modelica.Mechanics.Translational.Components at all. Might be also possible to keep the Vehicle example model and move the Vehicle component model to Modelica.Mechanics.Translational.Examples.Utilities to basically gain a new example model.

I have no strong preference here and would leave the decision to the libary officers @christiankral, @tobolar and @MartinOtter.

**Comment by christiankral** [comment]
> I believe there will be some general discussion if such a Vehicle model belongs to Modelica.Mechanics.Translational.Components at all. Might be also possible to keep the Vehicle example model and move the Vehicle component model to Modelica.Mechanics.Translational.Examples.Utilities to basically gain a new example model.
> 
> I have no strong preference here and would leave the decision to the libary officers @christiankral, @tobolar and @MartinOtter.

I am in favor of having the vehicle model under `Components` instead of `Utilities`. `Components` were the very first place where I would look at, if I were looking for the vehicle model.

**Comment by AHaumer** [comment]
@beutlich since I'm not familiar with the enhanced git processes:
How do I reset on my side to get these changes?

**Comment by beutlich** [comment]
In the git bash call:

```sh
git fetch --all
git reset --hard 2940A
```

**Review by tobolar — DISMISSED** [review]
To me, the `Vehicle` component would be fine now.

But there is another issue. Even if I understand the reason to add signal input to `SignForce`, there are two reasons to reject this. First, the documentation of that component is obsolete since it still says it is a "constant force" component. This could be simply fixed.

Second, if this input signal would be accepted, `ConstantForce` and (at least) `ForceStep` shall be changed accordingly in order to have consistent definition/usage of those three components. Then the component `Force` could be removed since it would be of the same functionality like `ConstantForce`.

So we probably need another solution.

Additionally, the usage of rotationalMultiSensor is questionable as it is not necessary for the model functionality. This sensor can be simply add outside of the model, if needed by user.

And I would prefer to use wheel radius as parameter instead of wheel diameter. The radius is anyway used in docu in the equation of equivalent additional mass as well as in idealRollingWheel instance.

**Comment by beutlich** [comment]
@AHaumer @christiankral Could you please care for the requested changes by the reviewers such that we can merge it finally?

**Comment by AHaumer** [comment]
Followed @tobolar 's review regarding the multiSensor and wheel diameter resp. radius.
I've to check another solution for rolling resistance.

**Comment by AHaumer** [comment]
@tobolar 
I understand that you prefer to have the sources consistent (also comparing Translational and Rotational). Anyhow, we need the rolling resistance dependent on rolling resistance coefficient and inclination and additionally there's the need for regularization around zero speed to exclude static friction (for simplicity). Would you accept a separate RollingResistance model (replacing the SignForce from which the input would be removed)? Placed in Sources or in Components?

**Comment by tobolar** [comment]
> we need the rolling resistance dependent on rolling resistance coefficient and inclination

I fully accept this, no question about it.

Having RollingResistance is a preferable solution for me. As this is a force element, it could be placed in Sources. But on the other side it not really fit the concept of that package. So either we can
- let it in Components, or
- define it nested in the `Vehicle` model.

@beutlich Any preference from the MSL design view?

Some comments on `RollingResistance`:
- the inputs `inclination` and `cr` may not necessarilly by conditional; but if this is desired for some reason then `usecrInput` and `useInclinationInput` of the instance `rollForce` in the `Vehicle` shall be propageted from the corresponding parameters (instead of setting them both to true only - then you need not to set `crConstant` and `inclinationConstant` here)
- description of the parameter `fNormal`: this force is generally called "wheel load" or "vertical force"
- `inclinationConstant` has a default value, but `crConstant` just a start value; I guess both can have default; see also `Modelica.UsersGuide.ParameterDefaults`
- even `v0`could heve a default value, cf. `w_small` in `Modelica.Mechanics.Rotational.Components.Clutch`

**Comment by AHaumer** [comment]
@tobolar 
You're right, Sources is not the right place for RollingResistance, it fits in Components. Please don't hide it in the Vehicle model.
I'll follow your hints regarding the inputs and the default values.

**Comment by AHaumer** [comment]
@tobolar - would you please have a look at OneWayClutch (not Clutch):
w_small=1e10 doesn't seem like a "Relative angular velocity near to zero ..." to me.

**Comment by tobolar** [comment]
> Would you please have a look at OneWayClutch (not Clutch).
w_small=1e10 doesn't seem like a "Relative angular velocity near to zero ..." to me.
Should this be a new ticket or am I wrong?

If any doubt, simply open a new ticket.

**Comment by HansOlsson** [comment]
> If any doubt, simply open a new ticket.

There is already one: #1493

**Comment by tobolar** [comment]
@AHaumer 
> - even v0 could have a default value, cf. w_small in Modelica.Mechanics.Rotational.Components.Clutch

My idea was not to set v0 to the value of w_small but to show some reference model where a regularization parameter (here w_small) has a default value as well. So your original value of v0 = 0.1 m/s is fine. Just not being a start value but default instead.

> For cr I have no reasonable default, therefore only a start value (strongly depends on material pairing).

[Rill](https://www.crcpress.com/Road-Vehicle-Dynamics-Fundamentals-and-Modeling/Rill/p/book/9781439838983) (formerly OTH :-) ) says: "Under normal conditions ... the rolling resistance for typical passenger car tires is in the range of 0.01 to 0.02."
So we can take an average: crConstant = 0.015 or the value you suggest as a start: 0.01.

> - description of the parameter fNormal: this force is generally called "wheel load" or "vertical force"

I have also found "vertical wheel load". IMO we can use this one.

**Comment by tobolar** [comment]
@AHaumer Since you use fDrag, fRoll and fGrav in the documentation I suggest to rename the corresponding instances (dragForce, rollForce and gravForce, resp.) accordingly.

File: Modelica/Mechanics/Translational.mo
Comment (christiankral) [inline Modelica/Mechanics/Translational.mo]: The variable name `cw` is a German thing. In English `cd` may be used, see [Wikipedia](https://en.wikipedia.org/wiki/Drag_coefficient). Possibly a native English speaking person can comment on this. 

@mtiller or @harmanpa Could you possibly comment on this?

File: Modelica/Mechanics/Translational.mo
Comment (christiankral) [inline Modelica/Mechanics/Translational.mo]: The name `fNormal` may not be a good choice as the normal force is calculated by the expression

`fNormal*cos(atan(inclination_internal))`

in line 3440. As I understand the model `fNormal` is the weight (mass * acceleration of gravity) = `m * g` only, see line 3577.

File: Modelica/Mechanics/Translational.mo
Comment (harmanpa) [reply Modelica/Mechanics/Translational.mo]: The commonly understand variable name would be Cd

**Comment by AHaumer** [comment]
I followed @tobolar 's review.
Regarding @christiankral 's review and @harmanpa 's comment, I changed the parameter names from cw to Cd and cr to Cr (to be consistent), as well as fNormal to fWeight "Wheel load due to gravity" in model RollingResistance.
See 6ee8640ff..c95c538c3

**Comment by AHaumer** [comment]
@tobolar regarding the question about w_small (OneWayClutch):
@HansOlsson is right, there's already a ticket #1493 
Don't know why this ticket was closed - I reopened it, asking you kindly to take care of that.

**Review by christiankral — APPROVED** [review]
Looks OK from my side.

File: Modelica/Mechanics/Translational.mo
Comment (tobolar) [reply Modelica/Mechanics/Translational.mo]: I have also found c_D in https://www.crcpress.com/Road-and-Off-Road-Vehicle-System-Dynamics-Handbook/Mastinu-Ploechl/p/book/9781138075290 (but originating from a German-speaking author anyway ;-) )

**Comment by tobolar** [comment]
When renaming parameter cw to Cd, also the input cr has been changed to Cr. This leads to a) breaking the recommendation 3 "Instance names" in `Modelica.UsersGuide.Conventions.ModelicaCode.Naming` (well, this is just a recommendation and thus not that bad) and b) to some discrepancies in the documentation (esp. of RollingResistance):
>The rolling resistance coeffcient _Cr_ is either constant (given by the parameter `CrConstant`) or prescribed by the input `Cr`. 

I suggest the following naming in the model:

1. _Cr_ to indicate in general the rolling resistance coeffcient.
2. `cr` for the real input. (so change it back to lower case)
3. Parameter `CrConstant` is fine.

**Comment by AHaumer** [comment]
@tobolar done by c9d6441a1..e09ff284f

**Comment by tobolar** [comment]
@AHaumer In the documentation of Vehicle there is:
> The rotational inertia J can be expressed as an equivalent additional mass mEq = J/R2, where R is the wheel radius, but only the vehicle mass influences rolling resistance and inclination resistance.

The relationshp between mEq and J for idealized wheel contact is clear but as the model doesn't deal with the acceleration only one can't simply transform J to mEq and add it to the vehicle's mass. This is even stated in the subclause. So what is the real benefit of this information for the user?

File: Modelica/Mechanics/Translational.mo
Comment (beutlich) [inline Modelica/Mechanics/Translational.mo]: @tobolar EllipseClosure is legal Modelica 3.4. Why was it removed here?

Comment (tobolar) [reply Modelica/Mechanics/Translational.mo]: Looking at the icon, the arc/ellipse is located between the two lines of the "road". It represents the inclination angle. As the two existing lines close the angle already, the additional closure of the ellipse would imply the lines being doubled. Just look at the following icon's cutoff (Dymola). 
grafik

Comment (beutlich) [reply Modelica/Mechanics/Translational.mo]: Ah, sorry. I was wrong here.

**Comment by beutlich** [comment]
@tobolar Is this ready to merge then? If yes, please approve!

**Comment by tobolar** [comment]
I wait for reply of @AHaumer to https://github.com/modelica/ModelicaStandardLibrary/pull/2963#issuecomment-565379766.

**Comment by AHaumer** [comment]
@tobolar 
What's the problem of pointing out something that is clear to you but maybe not to all users. 
Maybe you should add a remark that mEq is only valid for acceleration, not for rolling resistance.
Or if you prefer remove the remark about mEq totally, both solutions are ok for me.

**Comment by tobolar** [comment]
@AHaumer 
> What's the problem of pointing out something that is clear to you but maybe not to all users.

This is fine in general.
If we retain the information on mEq, we shall underline this is *not* applicable in this particular model. I will modify the documentation accordingly.

**Comment by beutlich** [comment]
>
> **Anton Haumer** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).

@AHaumer Your mail address anton.haumer@oth-regensburg.de is not associated with your GitHub account. If you do not want to add this address, you should avoid using that mail address for MSL contributions in future. I still can merge this PR (even w/o this CLA issue corrected).

**Comment by tobolar** [comment]
@AHaumer 
> fine for me

I suppose this concerns both my comment https://github.com/modelica/ModelicaStandardLibrary/pull/2963#issuecomment-572496344 and that of @beutlich https://github.com/modelica/ModelicaStandardLibrary/pull/2963#issuecomment-573417236.

**Review by tobolar — APPROVED** [review]

**Pull request merged**
